### DISCUSSION TITLE
Receiving events

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#2703] [User Support]
Hi @toots,

I'm interested in having external processes receive events from liquidsoap, ie onTrack, onFrame, metadata etc. I could make a web request, but having an open stream / socket would be ideal.

I found #2014 but file.open it creates a file by default and it is unclear how to create a named socket, ie /tmp/liqevents.sock

Can you provide a snippet which shows how to have liquidsoap create this socket so that my liq script could then write to it in the on_x events?

PS I tried having an external process create the socket (socat) and gave a+x permissions but liquidsoap still complained it couldn't open the socket with an error

Below is how a normal liq server socket looks like on my system:
```
srw------- 1 10000     10001       0 Oct 22 22:53 liq.sock
```

Here is what I have created with socat:
```
socat UNIX-LISTEN:/tmp/liqevents.sock,reuseaddr,fork TCP-LISTEN:11111
srwxrwxrwx 1 10000     10001       0 Oct 22 22:55 liqevents.sock
```

In the liq script:
```
fd = file.open(write=true, "/tmp/liqevents.sock")
```

When starting...
```
Error 14: Uncaught runtime error:
type: file,
message: "No such device or address in open(/tmp/liqevents.sock)\nBacktrace:\nRaised by primitive operation at Builtins_file                          s.(fun) in file \"builtins/builtins_files.ml\", line 203, characters 10-47\n",
stack: at /test.liq, line 13, char 5-49
```

Looks like liquidsoap thinks its a file and can't access it, similar to how if I try to cp the socket, bash tells me cannot open for reading - no such device or address.

Tried with the latest 2.1.2 rolling release:
```
Liquidsoap 2.1.3+git@bf6eca1f
Copyright (c) 2003-2022 Savonet team
Liquidsoap is open-source software, released under GNU General Public License.
```

**Comment by toots** [comment]
Hi! That's an interesting use-case.

From a POSIX/unix point of view, there should be no reason for the call to fail here as opposed to doing it with a regular files. My suspicion would be that the access rights are not allowing the liquidsoap process to access the file.

**Comment by toots** [comment]
Wait my bad, I can reproduce. Let me get back to you!

**Comment by toots** [comment]
Ok, unix/posix `openfile` call does not in fact support named sockets. I am adding the whole socket API in this PR: https://github.com/savonet/liquidsoap/pull/2722 which will include support for named sockets. However, this will have to be for the next major release.

Meanwhile, I'd suggest to use `process.run` with `socat`, which should allow you to communicate with the named socket via the local shell.
